### PR TITLE
feat(spanner): add support for Drop Database Protection

### PR DESCRIPTION
**Pull request by harshachinta**
This PR adds code changes and integration tests for Drop Database Protection feature.

### CODE REVIEW

File: spanner/integration_test.go [hunk 1]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+
+	_, dbPath, cleanup := prepareIntegrationTest(ctx, t, SessionPoolConfig{}, []string{})
+	defer cleanup()
+
+	// Check if enable_drop_protection is false by default for a newly created database
+	if database, _ := databaseAdmin.GetDatabase(ctx, &adminpb.GetDatabaseRequest{Name: dbPath}); database.GetEnableDropProtection() {
+		t.Fatalf("enable_drop_protection must be false by default for the DB %v", dbPath)
+	}
+
+	// Enable drop database protection to the testing database
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4347]:
```suggestion
	// Enable drop protection on the database
```

File: spanner/integration_test.go [hunk 2]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+		}}
+	opUpdate, err := databaseAdmin.UpdateDatabase(ctx, updateDatabaseReq)
+	if err != nil {
+		t.Fatalf("cannot enable drop db protection to testing DB %v: %v", dbPath, err)
+	}
+	if _, err := opUpdate.Wait(ctx); err != nil {
+		t.Fatalf("cannot enable drop db protection to testing DB %v: %v", dbPath, err)
+	}
+
+	// Check if enable_drop_protection is true for the testing database
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4364]:
```suggestion
	// Check if enable_drop_protection field is set to true for the database
```

File: spanner/integration_test.go [hunk 3]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+	if database, _ := databaseAdmin.GetDatabase(ctx, &adminpb.GetDatabaseRequest{Name: dbPath}); !database.GetEnableDropProtection() {
+		t.Fatalf("enable_drop_protection must be true for the DB %v", dbPath)
+	}
+
+	// Dropping the testing database which is drop protection enabled must throw an error
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4369]:
```suggestion
	// Dropping the database (with enabled drop protection) must throw an error
```

File: spanner/integration_test.go [hunk 4]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+	err = databaseAdmin.DropDatabase(ctx, &adminpb.DropDatabaseRequest{Database: dbPath})
+	if err == nil {
+		t.Fatalf("should fail to drop testing database as drop db protection is enabled for %v", dbPath)
+	}
+	if msg, ok := matchError(err, codes.FailedPrecondition, "enable_drop_protection"); !ok {
+		t.Fatal(msg)
+	}
+
+	// Deleting the instance which has database with drop protection enabled must throw an error
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4378]:
```suggestion
	// Deleting the instance (having the database with drop protection enabled) must throw an error
```

File: spanner/integration_test.go [hunk 5]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+	instanceName := fmt.Sprintf("projects/%v/instances/%v", testProjectID, testInstanceID)
+	err = instanceAdmin.DeleteInstance(ctx, &instancepb.DeleteInstanceRequest{Name: instanceName})
+	if err == nil {
+		t.Fatalf("should fail to drop instance %v as one of the database has drop db protection enabled", instanceName)
+	}
+	if msg, ok := matchError(err, codes.FailedPrecondition, "drop protection enabled"); !ok {
+		t.Fatal(msg)
+	}
+
+	// Disable drop database protection to the testing database
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4388]:
```suggestion
	// Disable drop database protection on the database
```

File: spanner/integration_test.go [hunk 6]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
...
+		}}
+	opUpdate, err = databaseAdmin.UpdateDatabase(ctx, updateDatabaseReq)
+	if err != nil {
+		t.Fatalf("cannot disable drop db protection to testing DB %v: %v", dbPath, err)
+	}
+	if _, err := opUpdate.Wait(ctx); err != nil {
+		t.Fatalf("cannot disable drop db protection to testing DB %v: %v", dbPath, err)
+	}
+
+	// Should drop database successfully as protection is disabled
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4405]:
```suggestion
	// Should drop the database successfully as drop protection is disabled
```

File: spanner/integration_test.go [hunk 7]
```diff
@@ -4328,6 +4329,85 @@ func TestIntegration_GFE_Latency(t *testing.T) {
 	DisableGfeLatencyAndHeaderMissingCountViews()
 }
 
+// TestIntegration_DropDatabaseProtection tests the drop database protection feature
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4332]: Can we add an integration test to check if a non-null value is populated for the enable_drop_protection field in ListDatabases call as well? This field will be back-populated for old databases as well.

File: spanner/integration_test.go (hunk 1 above)
Comment (harshachinta) [reply spanner/integration_test.go]: Updated

File: spanner/integration_test.go (hunk 2 above)
Comment (harshachinta) [reply spanner/integration_test.go]: Updated

File: spanner/integration_test.go (hunk 4 above)
Comment (harshachinta) [reply spanner/integration_test.go]: Updated

File: spanner/integration_test.go (hunk 6 above)
Comment (harshachinta) [reply spanner/integration_test.go]: Updated

File: spanner/integration_test.go (hunk 5 above)
Comment (harshachinta) [reply spanner/integration_test.go]: Updated

File: spanner/integration_test.go (hunk 7 above)
Comment (harshachinta) [reply spanner/integration_test.go:4332]: As `enable_drop_protection` field is a boolean field, in golang this automatically defaults to `false` and cannot have a `nil` value in it. Hence instead of testing `enable_drop_protection` field, I have added a test to verify `ListDatabases` returns `non-null` databases which by default indicates that `enable_drop_protection` field is set to either `true`/`false`.

File: spanner/integration_test.go [hunk 8]
```diff
@@ -4331,6 +4331,8 @@ func TestIntegration_GFE_Latency(t *testing.T) {
 
 // TestIntegration_DropDatabaseProtection tests the drop database protection feature
 func TestIntegration_DropDatabaseProtection(t *testing.T) {
+	skipEmulatorTest(t)
+	skipUnsupportedPGTest(t)
```
Comment (rajatbhatta) [inline spanner/integration_test.go:4335]: I think PG databases are supported by this feature?

Comment (harshachinta) [reply spanner/integration_test.go]: Yes. Removed skipping PG test for drop database protection

**Review by rahul2393 — APPROVED** [review]

**Review by rajatbhatta — APPROVED** [review]
LGTM

**Review by rahul2393 — CHANGES_REQUESTED** [review]
We don't need integration tests for admin feature.
